curation/scripts/qc_ref_genome.py

```diff
@@ -0,0 +1,257 @@
+import sys
+import pandas as pd
+import subprocess
+import requests
+from requests.adapters import HTTPAdapter, Retry
+import json
+import argparse
+import re
+from typing import TypedDict
+
+# Ensembl REST server config
+server = "https://rest.ensembl.org"
+ext_seq = "/sequence/region/human"
+headers = {"Content-Type": "application/json", "Accept": "application/json"}
+MAX_SEQUENCE_REQUEST_SIZE = 50
+MAX_VARIATION_REQUEST_SIZE = 10  # maximum is 200, although it causes some requests to return no result
+
+usecols = ("rsID", "chr_name", "chr_position", "effect_allele", "other_allele")
+
+
+class Variant(TypedDict):
+    rsID: str
+    chr: str
+    pos: int
+    alleles: list[str]
+
+
+class MappingResults(TypedDict):
+    match: int
+    mismatch: int
+    errors: list[str]
+
+
+def report(msg: str, submsgs: list[str] = []):
+    print('# ' + msg)
+    for submsg in submsgs:
+        print('#    - ' + submsg)
+
+
+def warn(msg: str):
+    sys.stderr.write(msg + "\n")
+
+
+compl_alleles = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}
+
+
+def flip_alleles(alleles: list[str]):
+    """Flip the given alleles for checking for the reverse strand"""
+    return [compl_alleles[a.upper()] for a in alleles]
+
+
+def post_request_with_retry(url, data, retry: int = 0):
+    """Send a POST request to the given URL. The request will be retried up to specified number of times
+    if the returned response is empty yet successful (happens with Ensembl Variation API)"""
+    s = requests.Session()
+    retries = Retry(total=5, backoff_factor=1, status_forcelist=[500, 502, 503, 504],
+                    allowed_methods=frozenset(['GET', 'POST']))
+    s.mount('http://', HTTPAdapter(max_retries=retries))
+    s.mount('https://', HTTPAdapter(max_retries=retries))
+    r = s.post(url, headers=headers, data=json.dumps(data))
+
+    if not r.ok:
+        r.raise_for_status()
+        quit()
+
+    result = r.json()
+    if retry and not result:  # Sometimes Ensembl returns an empty response (although status 200)
+        warn('Returned empty results. Retrying request (remaining attempts: %i)' % retry)
+        result = post_request_with_retry(url, data, retry - 1)
+
+    return result
+
+
+def get_variation_from_ensembl(rsids: list[str], ref_genome):
+    url = 'https://grch37.rest.ensembl.org/variation/human/' if ref_genome == '37' else 'https://rest.ensembl.org/variation/human/'
```

Comment:
Should we store the 2 Ensembl URLs in a dictionnary instead ?
e.g.:
```
servers = {
   '37': 'https://grch37.rest.ensembl.org',
   '38': 'https://rest.ensembl.org/'
}
```
The root url is also stored in `server` line 12